File: 149 Max Points on a Line/Python/Solution.py
from typing import Dict, List, Self, Set


class Solution:
    def maxPoints(self: Self, points: List[List[int]]) -> int:
        l: int = len(points)
        if l == 0:
            return 1
        
        lines: Dict[str, Set[str]] = {}
        m: int = 0

        for i in range(l):
            for j in range (i, l):
                p1: List[int] = points[i]
                [x1, y1] = [p1[0], p1[1]]
                p2: List[int] = points[j]
                [x2, y2] = [p2[0], p2[1]]

                k: str = ''
                b: str = ''
                if x1 == x2:
                    k = "undefined"
                    b = x1
                else:
                    k = (y2 - y1) / (x2 - x1)
                    b = y1 - k * x1
                line: str = f"{k},{b}"

                if line in lines:
                    lines[line].add(f"{x1},{y1}")
                    lines[line].add(f"{x2},{y2}")
                    m = max(m, len(lines[line]))
                else:
                    lines[line] = set()
                    lines[line].add(f"{x1},{y1}")
                    lines[line].add(f"{x2},{y2}")
                    m = max(m, len(lines[line]))

        return m

File: 149 Max Points on a Line/Java/Solution.java
public class Solution {
    public int maxPoints(int[][] points) {
        if (points.length == 1) {
            return 1;
        }

        int max = 2;
        for (int i = 0; i < points.length; i++) {
            int[] p1 = points[i];
            int x1 = p1[0];
            int y1 = p1[1];

            for (int j = i + 1; j < points.length; j++) {
                int[] p2 = points[j];
                int x2 = p2[0];
                int y2 = p2[1];

                int count = 2;
                if (x1 == x2) {
                    for (int l = 0; l < points.length; l++) {
                        if (l == i || l == j) {
                            continue;
                        }

                        int x3 = points[l][0];
                        if (x3 == x1) {
                            count++;
                        }
                    }
                }
                else if (y1 == y2) {
                    for (int l = 0; l < points.length; l++) {
                        if (l == i || l == j) {
                            continue;
                        }

                        int y3 = points[l][1];
                        if (y3 == y1) {
                            count++;
                        }
                    }
                }
                else {
                    double k = (double) (y2 - y1) / (x2 - x1);
                    double b = y1 - k * x1;

                    for (int l = 0; l < points.length; l++) {
                        if (l == i || l == j) {
                            continue;
                        }

                        int[] p3 = points[l];
                        int x3 = p3[0];
                        int y3 = p3[1];
                        double potentialY = k * x3 + b;

                        if (Math.abs(y3 - potentialY) <= 0.00001) {
                            count++;
                        }
                    }
                }

                max = Math.max(max, count);
            }
        }

        return max;
    }
}
